src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -1800,20 +1812,37 @@ public ImmutableArray<DocumentId> GetDocumentIdsWithFilePath(string? filePath)
         private ImmutableDictionary<ProjectId, CompilationTracker> CreateCompilationTrackerMap(ProjectId projectId, ProjectDependencyGraph dependencyGraph)
         {
             var builder = ImmutableDictionary.CreateBuilder<ProjectId, CompilationTracker>();
-            var dependencies = dependencyGraph.GetProjectsThatTransitivelyDependOnThisProject(projectId);
+            IEnumerable<ProjectId>? dependencies = null;
 
-            foreach (var projectIdAndTracker in _projectIdToTrackerMap)
+            foreach (var (id, tracker) in _projectIdToTrackerMap)
             {
-                var id = projectIdAndTracker.Key;
-                var tracker = projectIdAndTracker.Value;
-
                 if (!tracker.HasCompilation)
                 {
                     continue;
                 }
 
-                var canReuse = id == projectId || !dependencies.Contains(id);
-                builder.Add(id, canReuse ? tracker : tracker.Fork(tracker.ProjectState));
+                builder.Add(id, CanReuse() ? tracker : tracker.Fork(tracker.ProjectState));
+
+                // Returns true if 'tracker' can be reused for project 'id'
+                bool CanReuse()
+                {
+                    if (id == projectId)
+                        return true;
+
+                    // Check the dependency graph to see if project 'id' directly or transitively depends on 'projectId'.
+                    // If the information is not available, do not compute it.
+                    var forwardDependencies = dependencyGraph.TryGetProjectsThatThisProjectTransitivelyDependsOn(id);
```

Comment:
Rather than exposing this Try* method, are we better off just having a method which is "DoesATransitivelyDependOnB"? This brings up a few nice benefits:

1. You don't need the odd Try* method wart which exposes the internal laziness.
2. You can easily optimize it in other ways too and everybody gets the benefit. (i.e. if A just doesn't have dependencies at all, you know the answer as a shortcut. Or you can check the non-transitive ones first before invoking the computation....)
3. The implementation here is making an implicit assumption about the other calls in the system. Checking both our forward and reverse maps is a great idea, but the implicit assumption is that if we have to compute one, we should compute the reverse transitive map. If we just move to a boolean-returning method, then other code like https://github.com/dotnet/roslyn/blob/103ee6cb60b1a12b86f59b6ed407a99c90fa110d/src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs#L2175 can *also* be updated to return this method so the assumption is consistent for everyone.
4. It means you can get rid of your dependencies local since the deferring of that computation is now happening inside of ProjectDependencyGraph automatically. This might remove your need for the local function.